test/optimal_filter_validation.jl

```diff
@@ -0,0 +1,897 @@
+
+using Distributions
+using LinearAlgebra
+using FFTW
+using Random
+
+####################################################################
+####################################################################
+
+Random.seed!(123)
+
+#### Parameters for Covariance Function
+struct f_th
+    s::Float64
+    l::Float64
+end
+
+#### Parameters for Grid
+struct f_gr
+    p::Int ## Number of Grid Points Along x-axis is (p+1)
+    q::Int ## Number of Grid Points Along y-axis is (q+1)
+    P::Float64
+    Q::Float64 ## 2D-Field is rectangle [0,P]x[0,Q]
+end
+
+#### Parameters for Stations
+struct f_st
+  st_ij::Array{Int64,2} ## the position of the stations on the grid
+                        ## i.e station positions determined using 2 co-ordinates
+   st_p::Array{Int64,1} ## the no of grid point occupied by each station
+                        ## i.e station positions determined using 1 co-ordinate
+end
+
+####################################################################
+#### Example #######################################################
+
+th = f_th(1.0, 1.0)
+
+gr = f_gr(20, 20, 40.0, 40.0)
+
+###################################################
+#### Example ######################################
+
+p = gr.p; q = gr.q; P = gr.P; Q = gr.Q
+
+Dx = P/p; Dy = Q/q ## Grid Steps Along x and y Directions
+
+###################################################
+#### Example ######################################
+
+#### In real case Stations will be read from a file containing co-ordinates;
+#### here, I delect position of n2 Stations at random from all grid nodes
+
+n2 = 3; n1 = (p+1)*(q+1)
+
+grid_F = fill(0,(n1,2))
+
+#### the loop contructs 2-dim co-ordinates for all grid points
+for j in 0:q
+
+  for i in 0:p
+
+    grid_F[j*(p+1)+i+1,:] = [i, j]
+
+  end
+
+end
+####
+
+st_p = sample(1:n1,n2,replace=false)
+
+st_ij = grid_F[st_p,:]
+
+st = f_st(st_ij, st_p)
+
+###################################################
+#### Example ######################################
+
+#### I simulate some arbitrary values to be used as observations of height
+#### I give an example value to the standard deviation Sobs;
+#### I only consider a single time instance
+
+Yobs_t = rand(Normal(0,1), n2)
+#### the ordering in the observations is in tandem with the ordering
+#### of the stations in variable st
+
+Sobs = 0.01
+
+###################################################
+###################################################
+
+#### Example Specification of Covariance Function
+function R(x::Float64, y::Float64, th)
+
+  Cov = (th.s^2)*exp(-(abs(x)+abs(y))/(2.0*th.l))
+
+  return(Cov)
+
+end
+
+###################################################
+###################################################
+
+#### Specification of 'Extented' Covariance Function
+#### following Method/Notation in Dietrich & Newsam (96)
+function bar_R(x::Float64, y::Float64, th, gr)
+
+  P = gr.P; Q = gr.Q
+
+  bar_Cov = 0.0
+
+  if ( (0<=x) & (x<=P) & (0<=y) & (y<=Q) )
```

Comment:
```suggestion
  if 0 <= x <= P && 0 <= y <= Q
```

Reply:
I'm just going to leave this file as it is since it's Alex's code. I'm sure you'd have a lot of comments :)